### PR TITLE
[ Misc ] non-uniform quantization via `compressed-tensors` for `Llama`

### PR DESCRIPTION
**Pull request by robertgshaw2-redhat**
SUMMARY:
* add `compressed-tensors` integration to support for nonuniform quantization
* pipe `layer_name` though `llama.py` to enable selecting quantization scheme on layer-by-layer basis
* add `lm-eval` test case for an example non-uniform model

FOLLOW UP PRs:
* our config parsing is getting complicated. We should add `compressed-tensors` as a dependency
* expand non-uniform to the remaining models (currently only support llama, we will fail loudly but confusingly otherwise)

This enables us to run a model with `fp8` for some layers and `fp16` for other layers
```python
from vllm import LLM
# this model has fp16, fp8, and int8!
model = LLM("nm-testing/Meta-Llama-3-8B-Instruct-nonuniform-test")
```

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/model_executor/layers/quantization/compressed_tensors/utils.py [hunk 1]
```diff
@@ -76,25 +77,106 @@ class QuantizationArgs(BaseModel):
...
 
-def find_first_name_or_class_match(
-        name: str,
-        module: Module,
-        targets: Iterable[str],
-        check_contains: bool = False) -> Optional[str]:
+# fused_name: List[shard_name]
+_FUSED_LAYER_NAME_MAPPING = {
+    "qkv_proj": ["q_proj", "k_proj", "v_proj"],
+    "gate_up_proj": ["gate_proj", "up_proj"]
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/compressed_tensors/utils.py:83]: nit: Since this is based on model definitions, it would be nice to register and query this from the model architecture class.

File: vllm/model_executor/models/llama.py [hunk 2]
```diff
@@ -189,6 +197,7 @@ def __init__(
         attention_bias = getattr(config, "attention_bias", False) or getattr(
             config, "bias", False)
         self.self_attn = LlamaAttention(
+            parent_name=f"{parent_name}.self_attn",
```
Comment (mgoin) [inline vllm/model_executor/models/llama.py:200]: I think all these parent_name arguments should be layer_name. This doesn't make sense for this attention layer currently.

File: vllm/model_executor/models/llama.py [hunk 3]
```diff
@@ -129,12 +134,14 @@ def __init__(
...
             quant_config=quant_config,
+            layer_name=f"{parent_name}.qkv_proj",
         )
         self.o_proj = RowParallelLinear(
             input_size=self.total_num_heads * self.head_dim,
             output_size=hidden_size,
             bias=bias,
             quant_config=quant_config,
+            layer_name=f"{parent_name}.o_proj",
         )
```
Comment (mgoin) [inline vllm/model_executor/models/llama.py:145]: nit: It is worth looking into if we could do some python-fu to get "o_proj" from the name of the variable it is assigned to

Comment (youkaichao) [reply vllm/model_executor/models/llama.py:145]: I thought about it, but finally think it is not worthwhile. We also have module list assignment, where it is confusing which variable to use:

`self.layers = ModuleList([Layer() for i in range(5)]`

it is not possible to directly construct the name `layers.{i}` .

let's just concat strings with one additional keyword argument.

File: vllm/model_executor/layers/quantization/compressed_tensors/utils.py (hunk 1 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/layers/quantization/compressed_tensors/utils.py:83]: I will do this in the PR that applies the naming through all the models

File: vllm/model_executor/models/llama.py (hunk 2 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/models/llama.py]: updated per discussion with Kaichao

File: vllm/model_executor/models/utils.py [hunk 4]
```diff
@@ -65,7 +65,7 @@ def make_layers(
                                             get_pp_group().world_size)
     modules = torch.nn.ModuleList(
         [PPMissingLayer() for _ in range(start_layer)] +
-        [layer_fn() for _ in range(start_layer, end_layer)] +
```
Comment (youkaichao) [inline vllm/model_executor/models/utils.py:68]: add `prefix=""` for `make_layers`

File: vllm/model_executor/models/llama.py [hunk 5]
```diff
@@ -170,6 +176,7 @@ class LlamaDecoderLayer(nn.Module):
 
     def __init__(
         self,
+        prefix: str,
```
Comment (youkaichao) [inline vllm/model_executor/models/llama.py:179]: add default, and move it to the last arg

File: vllm/model_executor/models/llama.py [hunk 6]
```diff
@@ -129,12 +133,14 @@ def __init__(
             total_num_kv_heads=self.total_num_kv_heads,
             bias=bias,
             quant_config=quant_config,
+            layer_name=f"{prefix}.qkv_proj",
```
Comment (youkaichao) [inline vllm/model_executor/models/llama.py:136]: unify them to use `prefix`

Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/models/llama.py:136]: Are you sure?

This is the `Linear` which is the root. So its no longer a prefix but rather this is the final `layer_name`

File: vllm/model_executor/models/utils.py (hunk 4 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/models/utils.py:68]: okay sounds good

**Review by youkaichao — APPROVED** [review]
LGTM, let's change to `prefix` for the argument name.

File: vllm/model_executor/models/llama.py (hunk 6 above)
Comment (youkaichao) [reply vllm/model_executor/models/llama.py:136]: what's the difference?

File: vllm/model_executor/models/utils.py [hunk 7]
```diff
@@ -53,7 +53,9 @@ def __init__(self, *args, **kwargs):
 
 
 def make_layers(
-    num_hidden_layers: int, layer_fn: Callable[[], torch.nn.Module]
+    num_hidden_layers: int,
+    layer_fn: Callable[[], torch.nn.Module],
```
Comment (youkaichao) [inline vllm/model_executor/models/utils.py:57]: typing is not correct. you can create a proto for type checking:

```python
class LayerFn(Protocol):
    def __call__(
        self, prefix="",
    ) -> torch.nn.Module:
        ...
```

File: vllm/model_executor/models/gpt2.py [hunk 8]
```diff
@@ -186,7 +186,8 @@ def __init__(
         self.wpe = nn.Embedding(config.max_position_embeddings, self.embed_dim)
         self.start_layer, self.end_layer, self.h = make_layers(
             config.num_hidden_layers,
-            lambda: GPT2Block(config, cache_config, quant_config))
+            lambda prefix: GPT2Block(config, cache_config, quant_config),
+            prefix="")
```
Comment (youkaichao) [inline vllm/model_executor/models/gpt2.py:190]: update `GPT2Block`?

**Comment by youkaichao** [comment]
@robertgshaw2-neuralmagic please also merge the latest main and take care of the code in https://github.com/vllm-project/vllm/pull/6516 .

File: vllm/model_executor/models/llama.py (hunk 6 above)
Comment (comaniac) [reply vllm/model_executor/models/llama.py]: I also vote for `prefix`, because `Linear` is just another `torch.nn.module`. Unifying the naming might be more clear.

File: vllm/model_executor/models/llama.py [hunk 9]
```diff
@@ -268,9 +277,11 @@ def __init__(
...
         self.start_layer, self.end_layer, self.layers = make_layers(
             config.num_hidden_layers,
-            lambda: LlamaDecoderLayer(config=config,
-                                      cache_config=cache_config,
-                                      quant_config=quant_config))
+            lambda prefix: LlamaDecoderLayer(config=config,
+                                             cache_config=cache_config,
+                                             quant_config=quant_config,
+                                             prefix=prefix),
+            prefix="model.layers")
```
Comment (youkaichao) [inline vllm/model_executor/models/llama.py:284]: why not use the `f"{prefix}.layers"`?

File: vllm/model_executor/models/mixtral.py [hunk 10]
```diff
@@ -258,8 +267,11 @@ def __init__(
         )
 
         self.start_layer, self.end_layer, self.layers = make_layers(
-            config.num_hidden_layers, lambda: MixtralDecoderLayer(
-                config, cache_config, quant_config=quant_config))
+            config.num_hidden_layers,
+            lambda prefix: MixtralDecoderLayer(
+                config, cache_config, quant_config=quant_config, prefix=prefix
+            ),
+            prefix="model.layers")
```
Comment (youkaichao) [inline vllm/model_executor/models/mixtral.py:274]: same here.

Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/models/mixtral.py:274]: Sure ill back it out one layer higher
